### PR TITLE
S4748611 StyleGAN for ADNI Dataset with TSNE Embedding

### PR DESCRIPTION
**Pull request by SamWolfenden**
Hi Teaching Team,
This is my pull request for my implementation of a StyleGAN model for task 8 using Tensorflow. This model uses a Generative Adversarial Network with and incorporated Adaptive Instance Normalisation layer to synthesize high quality images of both “Normal Cognition” and “Alzheimer’s Disease” brains from the Alzheimer’s Disease Neuroimaging Initiative (ADNI) dataset. For more information about the training process, model architecture and generated images see the detailed README. 
The project contains the following:

- 	README.md is a report of the entire project, including the model implementation, training information, and images of the 
        output from the models including sample output images of the brains and a TSNE embedding.
- train.py contains the code required to train and save the StyleGAN model
- 	modules.py contains the code required to build the StyleGAN model.

- 	dataset.py contains the code required to load the ADNI dataset.

- 	predict.py contains the code required to generate new samples from the model created by the train file.

- 	Images folder contains all images used in the readme to highlight the StyleGAN model in use

Thankyou,
Sam Wolfenden
s4748611

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (4.5 Marks)
Reasonably clear generated images;
Losses of the discriminator are unusual but left unexplained -0.5;

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Modularized design

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- Nice GIFs showing progressive training outputs;
- Neat module structure;
- The losses of discriminator are unusually increasing while the losses of generator are dropping, this can suggest the training needs to be improved to achieve the equilibrium and consequently better generated images;
- You essentially had two trainers to train the same model on two classes of data, with embedding on two style codes plotted respectively. This is fine and I'm not taking marks off given the generated images are "reasonably clear" and the embedding is very clustered. I just want to point out the intention of this task (see the reference paper) is actually to train one model on the whole dataset, and by plotting the embedding of style code, it is expected to see somehow clustered patterns for different classes.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Nice documentation. T-SNE presents clear two clusters. Could have more introduction of the network. Nice visualization. Loss plot titles are unclear, what are the hyperparameters of green and purple parts? The generated images are ok.
2. **Implementation**: Some hyperparameters were hard-code; otherwise, nice implementations of functions and classes.
3. **Design**:
4. **Comment**: More detailed comments are required, like input and output
5. **Difficulty**: Hard

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. High losses.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Some/Adequate Meaningful commit messages. Add more discriptive messages.</td><td>-1</td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-3</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Meanginful commits -1
